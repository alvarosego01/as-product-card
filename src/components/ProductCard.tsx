

import { useProducts } from "../hooks/useProducts";
import React, { createContext } from "react";
import { InitialValues_I, ProductCardHandlers_I, ProductContextProps_I, Product_I, onChangeArgs_I } from "../interfaces/interfaces";

import styles from "../styles/styles.module.css";


export const ProductContext = createContext({

} as ProductContextProps_I)
const { Provider } = ProductContext;


export interface Props_I {
    product: Product_I;
    // children?: ReactElement | ReactElement[];
    children: (args: ProductCardHandlers_I) => JSX.Element;
    className?: string;
    style?: React.CSSProperties;
    onChange?: (args: onChangeArgs_I) => void;
    value?: number;
    initialValues?: InitialValues_I
}


export const ProductCard = ({ children, product, className, style, onChange, value, initialValues }: Props_I) => {

    const {
        counter,
        increaseBy,
        maxCount,
        isMaxCountReached,
        reset
    } = useProducts({
        onChange,
        product,
        value,
        initialValues
    });

    return (
        // Se provee la informaciÂ´pn que necesitan los hijos
        <Provider value={{
            counter,
            increaseBy,
            product,
            maxCount,
        }}>
            <div style={style} className={`${styles.productCard} ${className}`}>
                {
                    children({
                        count: counter,
                        isMaxCountReached,
                        maxCount,
                        product,
                        increaseBy,
                        reset
                    })
                }
            </div>
        </Provider>
    )

}


