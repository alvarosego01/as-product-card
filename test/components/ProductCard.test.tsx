import React from 'react';
import ProductCard from '../../src/components';
// import { render } from 'react-dom';
// import { render, screen } from "@testing-library/react";

import renderer from 'react-test-renderer';

import { product1 } from '../data/products';

const {act} = renderer;

describe('Pruebas en productTitle', () => {

    test('Debe mostrar component con titulo personalizado', () => {

        const wrapper = renderer.create(
         <ProductCard product={product1} >
                {
                    () => (
                       <h1>Product card</h1>
                    )
                }
            </ProductCard>
        );

        // console.log('wrapper', wrapper.toJSON());

        expect(wrapper.toJSON()).toMatchSnapshot();


    });

    test('Debe incrementar el contador', () => {

          const wrapper = renderer.create(
         <ProductCard product={product1} >
                {
                    ({count, increaseBy}) => (
                        <>
                             <h1>Product card</h1>
                            <span>{count}</span>
                            <button onClick={() => increaseBy(1)}>

                            </button>
                        </>
                    )
                }
            </ProductCard>
        );

        let tree = wrapper.toJSON();
        expect(tree).toMatchSnapshot();

        act(() => {
        (tree as any).children[2].props.onClick();
        })

        tree = wrapper.toJSON();

        expect((tree as any).children[1].children[0]).toBe("1");


     })



});
