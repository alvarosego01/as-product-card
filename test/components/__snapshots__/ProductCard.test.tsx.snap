// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en productTitle Debe incrementar el contador 1`] = `
<div
  className="productCard undefined"
>
  <h1>
    Product card
  </h1>
  <span>
    0
  </span>
  <button
    onClick={[Function]}
  />
</div>
`;

exports[`Pruebas en productTitle Debe mostrar component con titulo personalizado 1`] = `
<div
  className="productCard undefined"
>
  <h1>
    Product card
  </h1>
</div>
`;
